### DISCUSSION TITLE
Rethinking the sync algorithm

### DISCUSSION THREAD

**Discussion by HerbCaudill** [HerbCaudill/crdx#8] [Ideas]
The sync process in crdx uses the Bloom filter approach from [this paper](https://arxiv.org/pdf/2012.00472) by Martin Kleppmann (@ept) and Heidi Howard, just like Automerge. 

With links encrypted, I'm having to revisit some elements of the process, having to do with the fact that we don't have the references to previous heads in cleartext. Since Bloom filters are probabilistic, you don't always know what you don't know — Alice might think that Bob has a link when in fact he doesn't, and the only way for Bob to realize that is to try to reconstruct the chain and realize that there are gaps. But with encrypted links, we can't do that; and we can't decrypt the chain if there are gaps. 

The [argument for using Bloom filters](https://martin.kleppmann.com/2020/12/02/bloom-filter-hash-graph-sync.html) rather than just sending over hashes is that it's a very compact way of communicating what we know — you wouldn't want to have to send over tens or hundreds of thousands of hashes every time you wanted to sync up with someone. 

But the thing about signature chain is that all the activity happens at the head of the chain. If Alice and Bob are syncing up and they've diverged, it doesn't follow that they have to tell each other _every single_ hash they know about: They can start with the most recent ones, and send more if they still haven't found the branching point. 

So I think it would make more real-world sense to handle divergent heads by exchanging hashes of the first N links working backwards from the current heads. (This is what `git fetch` does, with N=32.) Even though the working-offline-for-a-long-time scenario looms large in our minds, I think it's reasonable to assume that divergence is more likely to be short-lived than long-lived, meaning that the point of divergence is likely to be found in the first N ancestors of our current head(s). 

That N value might be 10 or 100 or even 1000, depending on the chattiness and granularity of operations in the application; but it's constant — it doesn't grow as the dataset grows. 

Even better, we could include dependency information relatively compactly — so rather than just say that we know hashes A, B, C, and D, say that A depends on B, and B depends on C and D. This way we can easily figure out whether we have enough information about recent hashes, or if we need to ask for more.

**Comment by ept** [comment]
Hi Herb, if you are expecting the chain to grow fairly slowly, then simply exchanging the most recent N hashes (and working backwards until you reach a common ancestor) seems fine, and it has the advantage of being a simpler protocol. The Bloom filter protocol was intended to deal with scenarios where the chain grows quickly, e.g. in a text editor where every keystroke is a change, and so an offline editing session can easily accumulate tens of thousands of changes. But I guess that's not what crdx is designed for, so you can keep things simpler.

One thing you could also consider is to leave the hashes in the chain unencrypted, since they aren't really sensitive information. However, if you do this, you probably do need to authenticate the hashes, for example by passing them as the "additional data" of an [AEAD encryption scheme](https://en.wikipedia.org/wiki/Authenticated_encryption#Authenticated_encryption_with_associated_data_(AEAD)), to stop an adversary from modifying the hashes. Annoyingly, libsodium's `crypto_box` API doesn't seem to expose an additional data parameter, but the lower-level ChaCha20-Poly1305 APIs do.

**Reply by HerbCaudill** [reply]
> you probably do need to authenticate the hashes ... to stop an adversary from modifying the hashes

I was wondering about this, but I can't think of how an attack like this would work, in order to create a failing test. 

Links are keyed using the hash of the encrypted body. So if I take a link, modify it, and then modify the hash as well, I'm basically introducing a new link rather than replacing the existing one. There's no such thing as deleting an existing link. And a new link — well, if I'm syncing with you, I can give you any number of new links, but they'd have to be decrypted, authenticated, and validated in order to be accepted onto the chain. To do that I'd either need to have the appropriate permissions myself, or have someone else's credentials.

**Reply by ept** [reply]
As long as the hash of a node in the chain includes the parent hashes, and the parent hashes are authenticated using a signature, then it should be fine. I just wasn't sure what your construction looks like exactly.

**Comment by HerbCaudill** [comment]
This work is mostly done, on the `encrypt-chain` branch — see #6 .

I started out implementing this as described, using the `git-fetch` algorithm where we send the hashes and dependencies up to N levels of depth, and ask for more if needed. But our sync state with each peer includes `lastCommonHead`, so we actually have an upper limit for how far back we need to go when sending them hashes; so we might as well send them all at once and be 100% assured that they will then have all the information they need to request the links they need to be synced up.

We may eventually see that some sync messages are too big and need to be broken up into multiple messages, but we'll see that first in scenarios where we're sending the actual links, not when sending hashes. Either way we can cross that bridge when we get there.
